### PR TITLE
Add symfony dependency injection container [MAILPOET-1605]

### PR DESCRIPTION
**Pull request by costasovo**
# Mozart Plugin Update
Mozart plugin is used to move composer packages classes to our namespace to avoid conflicts with other plugins.
In this PR I've moved configuration of dependencies managed by Mozart to an extra configuration file `mozart/composer.json` and added its processing into main `composer.json` scripts and also to `build.sh`
This was done because we want to delete dependencies after they are moved to a new namespace so that we are safe from making a mistake by using unprefixed class. Some deps were also locked by dev dependencies.

# Symfony DI container
I've installed a minimum of Symfony dependencies so that we are able to use container and keep mailpoet.zip reasonable size.
This is the reason why I used basic PHP configuration instead of more comfortable `services.yaml` file.

## Basics
A container is created and configured in `libs/DI/ContaineFactory.php`. The easiest way and also preferred way of adding a service to this container is to pass all dependencies to the service by a constructor with proper type hints. Such a service can be registered easily using autowiring `$container->autowire(\MailPoet\YourService::class);`.
We use autowire to keep DI Configuration small. Later we will be make configuration even smaller by [registering whole namespaces](https://symfony.com/doc/3.4/service_container.html#multiple-service-definitions-using-the-same-namespace)

Apart of services you can register also parametes:
`$container->setParameter('mailpoet.assets_url', Env::$assets_url);`

If you want to define a service which requires a parameter as dependency you can still use autowire and define just the parameter with reference to its variable name: `$container->autowire(ServiceWhichNeedAssetsURl::class)->setArgument('$assets_url', '%mailpoet.assets_url%');` If the service has other dependencies they will be autowired.

## Compilation
In order to provide autowired services the container has to be compiled (`$container->compile()`). This operation can be time-consuming so that's why we store compiled container in `libs/DI/CachedContainer.php`. This cache is ignored when `WP_DEBUG` is enabled or when the cache file is not present.

I have added a command `./do container:dump` and also a build step to ./build.sh. Since some classes check for `ABSPATH` constant in header (I guess it is some security practice to ensure php is executed in WP environment) and some even require files from WordPress I've renamed `WP_TEST_PATH` and `WP_TEST_PATH_MULTISITE` variables to  `WP_ROOT` and `WP_ROOT_MULTISITE`to our `.env` file. You have to set the path to your WordPress root directory in order to be able to run `./do container:dump`

**Please update your .env file and also rebuild acceptance tests docker `docker-compose build codeception`**

### Before
```
class Dependency {

    /** @var array */
    private $params;

    function __construct($params) {
        $this->params = $params;
    }

    function do() {
        call($this->params, Env::api_key);
    }
}

class Service {
    function run($params) {
        $dep = new Dependency($params);
        $dep->do();
    }
}

// Main
$service = new Service();
$service->run(['x' => 'y']);
```

### After

```
class Dependency {

    /** @var string */
    private $api_key;

    function __construct($api_key) {
        $this->api_key = $api_kye;
    }

    function do($params) {
       call($params, $this->api_key);
    }
}

class Service {

    /** @var Dependency */
    private $dependency;

    function __construct(Dependency $dependency) {
        $this->dependency = $dependency;
    }

    function run($params) {
        $this->dependency->do($params);
    }
}

// Configuration inside factory class
$container->setParameter('service_api_key', '123456');
$container->autowire(Service::class);
$container->autowire(Dependency::class)->setArgument('$api_key', '%service_api_key%');

// Main
$service = $container->get(Service::class);
$service->run(['x' => 'y']);
```

- `Service` is no longer responsible for creating `Dependency` sub-service and of obtaining API key for it.
- `Service` can now be easily provided (only change in typehint) with an alternative `DependencyV2` which may have much more complicated construction. (This can be taken even further by using [interfaces and interface bindings](https://symfony.com/blog/new-in-symfony-3-4-local-service-binding))
- The number of dependencies in the constructor is a great indicator of the class complexity. We can easily notice e.g. during code review that class is getting too complex and should be divided.
- We have removed the hidden dependency on `Env` class from Service.
- If we add another dependency into service constructor we have to change only the constructor
- Dependency can be easily mocked and passed to the constructor in tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,10 +96,23 @@ $ ./do delete:docker      # stop and remove all running docker containers.
 $ ./do qa:lint             # PHP code linter.
 $ ./do qa:lint:javascript  # JS code linter.
 $ ./do qa                  # PHP and JS linters.
+
+$ ./do container:dump      # Dumps DI container.
```
Comment (pavel-mailpoet) [inline README.md]: This is not very helpful. The second part `# Dumps DI container.` doesn't say anything above the first part and doesn't tell us why should we dump the container.

File: RoboFile.php [hunk 2]
```diff
@@ -282,6 +282,21 @@ function testFailedIntegration() {
     return $this->_exec('vendor/bin/codecept run integration -g failed');
   }
 
+  function containerDump() {
+    $this->say('Deleting DI Container');
+    $this->_exec('rm ./lib/DI/CachedContainer.php');
```
Comment (pavel-mailpoet) [inline RoboFile.php]: This step can end up with an error if the file is not present. Could we suppress the error? It looks like the whole command failed.

File: build.sh [hunk 3]
```diff
@@ -22,11 +22,19 @@ test -d node_modules && rm -rf node_modules
 npm install
 ./do compile:all --env production
 
+# Dependency injection container dump.
+echo '[BUILD] Dumping dependency injection container'
```
Comment (pavel-mailpoet) [inline build.sh]: Again: `Dumping dependency injection container` doesn't say anything. Should it be something like: `Creating dependency injection container cache`?

File: README.md [hunk 4]
```diff
@@ -94,10 +96,23 @@ $ ./do delete:docker      # stop and remove all running docker containers.
...
 
+## DI
+
+We use Symfony/dependency-injection container. Container configuration can be found in `libs/DI/ContainerFactory.php`
+The container is configured and used with minimum sub-dependencies to keep final package size small. 
+You can check [the docs](https://symfony.com/doc/3.4/components/dependency_injection.html) to learn more about Symfony Container.
+
+## Mozart
+
+We use Mozart plugin for composer to prevent plugin libraries conflicts in PHP. Mozart prefix dependencies namespaces and moves them into `libs\Dependencies` directory.
```
Comment (pavel-mailpoet) [inline README.md]: Should we also mention how could the conflicts happen? Something about multiple WP plugins using the same composer dependency with different versions?

**Comment by costasovo** [comment]
I've updated these comments.

**Pull request merged**